Upgrade 9_4_X to Herwig 7.2

Dominic-Stafford opened the pull request:
#### PR description:

Adds Herwig 7.2.2 to CMSSW_9_4_X (replacing Herwigpp) for use in the Heavy Ion run (requested by the top group). Also adds Herwig Settings blocks, and workflow 535 for testing. Follow up to #28649, which was not merged due to a build error (possibly because the old ThePEGInterface was not removed)

#### PR validation:

Tested with workflow 535, and matchbox test command

#### if this PR is a backport please specify the original PR and why you need to backport that PR:
Backports a large number of PRs on the master branch upgrading Herwig

Dominic-Stafford committed:
Add Herwig7Interface, remove old ThePEGInterface, add Herwig 7 Settings, and add workflow 535 for testing

cmsbuild commented:
A new Pull Request was created by  @Dominic-Stafford  for CMSSW_9_4_X.

It involves the following packages:

Configuration/Generator
Configuration/PyReleaseValidation
GeneratorInterface/Herwig7Interface
GeneratorInterface/ThePEGInterface


@SiewYan, @mkirsano, @jordan-martins, @chayanit, @wajidalikhan, @kpedro88, @cmsbuild, @GurpreetSinghChahal, @srimanob, @agrohsje, @alberto-sanchez can you please review it and eventually sign? Thanks.
@mkirsano, @makortel, @Martin-Grunewald, @alberto-sanchez, @fabiocos, @slomeo this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

mrodozov commented:
please test with cms-sw/cmssw#33492

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-947a95/14392/summary.html
**COMMIT**: 9a51ea8794f3a487778b2a704d0a5b7a054362dc
**CMSSW**: CMSSW_9_4_X_2021-04-18-0000/slc6_amd64_gcc630
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/33492/14392/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Entering Package GeneratorInterface/Herwig7Interface
Entering library rule at src/GeneratorInterface/Herwig7Interface/plugins
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/plugins/Herwig7Hadronizer.cc 
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/interface/Herwig7Interface.h:27:0,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/plugins/Herwig7Hadronizer.cc:27:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/interface/HerwigUIProvider.h:7:33: fatal error: Herwig/API/HerwigUI.h: No such file or directory
 #include <Herwig/API/HerwigUI.h>
                                 ^
compilation terminated.
gmake: *** [tmp/slc6_amd64_gcc630/src/GeneratorInterface/Herwig7Interface/plugins/GeneratorInterfaceHerwig7HadronizerPlugins/Herwig7Hadronizer.o] Error 1
>> Building edm plugin tmp/slc6_amd64_gcc630/src/GeneratorInterface/Herwig7Interface/plugins/GeneratorInterfaceHerwig7HadronizerPlugins/libGeneratorInterfaceHerwig7HadronizerPlugins.so
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/src/RandomEngineGlue.cc 
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/src/HerwigUIProvider.cc 
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/src/Proxy.cc 
>> Compiling  /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/src/Herwig7Interface.cc 
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/src/HerwigUIProvider.cc:6:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/interface/HerwigUIProvider.h:7:10: fatal error: 'Herwig/API/HerwigUI.h' file not found
#include <Herwig/API/HerwigUI.h>
         ^~~~~~~~~~~~~~~~~~~~~~~
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-18-0000/src/GeneratorInterface/Herwig7Interface/src/Herwig7Interface.cc:24:10: fatal error: 'Herwig/API/HerwigAPI.h' file not found
#include <Herwig/API/HerwigAPI.h>
         ^~~~~~~~~~~~~~~~~~~~~~~~
</pre>

Dominic-Stafford committed:
Change herwigpp to herwig7 in BuildFile

cmsbuild commented:
Pull request #33492 was updated. @SiewYan, @mkirsano, @jordan-martins, @chayanit, @wajidalikhan, @kpedro88, @cmsbuild, @GurpreetSinghChahal, @srimanob, @agrohsje, @alberto-sanchez can you please check and sign again.

Dominic-Stafford commented:
Sorry, I missed changing "herwigpp" to "herwig7" in the build file for the plugins. Hopefully it should compile now

mrodozov commented:
allow @Dominic-Stafford test rights

mrodozov commented:
lets check if this will work because I think you'll have headers troubles

mrodozov commented:
yup now you can test it, try it

Dominic-Stafford commented:
please test with #33492

Dominic-Stafford commented:
Thanks. I built it just now on lxplus, so hopefully it should work, but we'll see

cmsbuild commented:
-1

**Failed Tests**:   Build  ClangBuild 
**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-947a95/14416/summary.html
**COMMIT**: 998b96f1771cc3041ed14075cf71fff335e064e0
**CMSSW**: CMSSW_9_4_X_2021-04-22-1100/slc6_amd64_gcc630
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/33492/14416/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Build

I found compilation error when building: 

<pre>>> Entering Package GeneratorInterface/Herwig7Interface
Entering library rule at src/GeneratorInterface/Herwig7Interface/plugins
>> Compiling edm plugin /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-22-1100/src/GeneratorInterface/Herwig7Interface/plugins/Herwig7Hadronizer.cc 
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-22-1100/src/GeneratorInterface/Herwig7Interface/interface/Herwig7Interface.h:27:0,
                 from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-22-1100/src/GeneratorInterface/Herwig7Interface/plugins/Herwig7Hadronizer.cc:27:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-22-1100/src/GeneratorInterface/Herwig7Interface/interface/HerwigUIProvider.h:7:33: fatal error: Herwig/API/HerwigUI.h: No such file or directory
 #include <Herwig/API/HerwigUI.h>
                                 ^
compilation terminated.
gmake: *** [tmp/slc6_amd64_gcc630/src/GeneratorInterface/Herwig7Interface/plugins/GeneratorInterfaceHerwig7HadronizerPlugins/Herwig7Hadronizer.o] Error 1
>> Building edm plugin tmp/slc6_amd64_gcc630/src/GeneratorInterface/Herwig7Interface/plugins/GeneratorInterfaceHerwig7HadronizerPlugins/libGeneratorInterfaceHerwig7HadronizerPlugins.so
</pre>

## Clang Build

I found compilation error while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```


<pre>  class DoNotRecordParents;
  ^~~~~
  struct
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-22-1100/src/GeneratorInterface/Herwig7Interface/plugins/Herwig7Hadronizer.cc:27:
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-22-1100/src/GeneratorInterface/Herwig7Interface/interface/Herwig7Interface.h:27:
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-22-1100/src/GeneratorInterface/Herwig7Interface/interface/HerwigUIProvider.h:7:10: fatal error: 'Herwig/API/HerwigUI.h' file not found
#include <Herwig/API/HerwigUI.h>
         ^~~~~~~~~~~~~~~~~~~~~~~
1 warning and 1 error generated.
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-22-1100/src/GeneratorInterface/Herwig7Interface/plugins/Herwig7Hadronizer.cc:27:
In file included from /data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_9_4_X_2021-04-22-1100/src/GeneratorInterface/Herwig7Interface/interface/Herwig7Interface.h:27:
</pre>

Dominic-Stafford commented:
Please test with cms-sw/cmsdist#6829

mrodozov commented:
abort

Dominic-Stafford commented:
Was there a reason these didn't start? I think it should work with this test command (sorry for using the wrong one initially)

mrodozov commented:
please test with cms-sw/cmsdist#6829

mrodozov commented:
sorry for this delay, yesterday at some point multiple prs failed in here, cmsdist and bot with the server aborting the jobs.
not sure what happened and unless it happens again I'll just restart the jobs

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-947a95/14541/summary.html
**COMMIT**: 998b96f1771cc3041ed14075cf71fff335e064e0
**CMSSW**: CMSSW_9_4_X_2021-04-22-1100/slc6_amd64_gcc630
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week1/cms-sw/cmssw/33492/14541/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 27
 - DQMHistoTests: Total histograms compared: 2721493
 - DQMHistoTests: Total failures: 108
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2721223
 - DQMHistoTests: Total skipped: 162
 - DQMHistoTests: Total Missing objects: 0
 - [TriggerResults](https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_9_4_X_2021-04-22-1100+947a95/42379/triggerResults): no differences found

agrohsje commented:
Hi @Dominic-Stafford @theofil : Do we need more workflows for validation? You changed 55 files. Not all changes would be caught by 535 and the above checks. Should we extend or merge for now and extend later the validation of H7. Let me know.

Dominic-Stafford commented:
Hi @agrohsje. In this case I changed 55 files because I'm backporting the entire Herwig 7 Interface, however I wouldn't expect this to cause problems because it's the same code as in the master branch. In addition to testing with workflow 535, I also test all the PRs I make with workflow 511, which tests the matchbox, but is currently not in the list of standard tests because it takes to long, I believe.

In general if we wanted to add more workflows we could add one or both of these tests I added recently:
https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/Herwig7Interface/python/DYToll01234Jets_5f_LO_MLM_Madgraph_Herwig_13TeV_cff.py
https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/Herwig7Interface/python/DYToll012Jets_5f_NLO_FXFX_Madgraph_Herwig_13TeV_cff.py
The first of these tests madgraph with MLM merging, while the second is FxFx (though this does also take a bit of time). If some people from the computing side could comment on how much extra time we could spend on Herwig7 workflows this would be helpful.

theofil commented:
hi, regarding the extra validation workflows, would it make sense to make
some trivial gridpack e.g., with DY and 0 jets and no merging just for the
purpose to be faster and validate the Herwig + externalLHE from Madgraph ?
is there such a trivial gridpack already existing, or we need to produce
one ?

On Wed, Apr 28, 2021 at 12:07 PM Dominic-Stafford ***@***.***>
wrote:

> Hi @agrohsje <https://github.com/agrohsje>. In this case I changed 55
> files because I'm backporting the entire Herwig 7 Interface, however I
> wouldn't expect this to cause problems because it's the same code as in the
> master branch. In addition to testing with workflow 535, I also test all
> the PRs I make with workflow 511, which tests the matchbox, but is
> currently not in the list of standard tests because it takes to long, I
> believe.
>
> In general if we wanted to add more workflows we could add one or both of
> these tests I added recently:
>
> https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/Herwig7Interface/python/DYToll01234Jets_5f_LO_MLM_Madgraph_Herwig_13TeV_cff.py
>
> https://github.com/cms-sw/cmssw/blob/master/GeneratorInterface/Herwig7Interface/python/DYToll012Jets_5f_NLO_FXFX_Madgraph_Herwig_13TeV_cff.py
> The first of these tests madgraph with MLM merging, while the second is
> FxFx (though this does also take a bit of time). If some people from the
> computing side could comment on how much extra time we could spend on
> Herwig7 workflows this would be helpful.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cms-sw/cmssw/pull/33492#issuecomment-828331891>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABDV54R5T5FRG3IKR2XYVWTTK7M7ZANCNFSM43K3I32A>
> .
>

agrohsje commented:
@chayanit: what do you think about adding mlm and fxfx herwig workflows. @theofil i think for mlm we can recycle the P8 gridpacks, for fxfx we need to rerun. But for nlo 2 jets is usually not so troublesome.

agrohsje commented:
+generators

chayanit commented:
+1 

I have no objection to add more workflows.

srimanob commented on Configuration/PyReleaseValidation/python/relval_steps.py:
```diff
@@ -950,6 +950,9 @@ def identityFS(wf):
                          },
                         step1Defaults])
 
+step1LHEGenDQM = merge([{'-s':'LHE,GEN,VALIDATION:genvalid','--datatier'    : 'GEN-SIM,LHE,DQMIO','--eventcontent': 'LHE,RAWSIM,DQM'},step1GenDefaults])
```
Should the datatier and eventcontent be in the same order?

srimanob replied:
See in master, for example,
https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_steps.py#L1435

Dominic-Stafford replied:
I don't think the order matters, as it's this way round in 9_3, which has run fine for a while, but I can change this if it does matter

srimanob replied:
Of course, It matters when you start to use it with relvals. You will register the mismatched dataset between eventcontent and datatier. See output definition from the PR test,
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-947a95/14694/runTheMatrix-results/535.0_TTbar_13TeV_Pow_herwig7+TTbar_13TeV_Pow_herwig7+HARVESTGEN/TT_13TeV_Pow_Herwig7_cff_LHE_GEN_VALIDATION.py

You will end up with `/..../..../LHE` with RAWSIM content and vice versa.

```
# Output definition

process.LHEoutput = cms.OutputModule("PoolOutputModule",
    dataset = cms.untracked.PSet(
        dataTier = cms.untracked.string('GEN-SIM'),
        filterName = cms.untracked.string('')
    ),
    fileName = cms.untracked.string('file:step1.root'),
    outputCommands = process.LHEEventContent.outputCommands,
    splitLevel = cms.untracked.int32(0)
)

process.RAWSIMoutput = cms.OutputModule("PoolOutputModule",
    SelectEvents = cms.untracked.PSet(
        SelectEvents = cms.vstring('generation_step')
    ),
    compressionAlgorithm = cms.untracked.string('LZMA'),
    compressionLevel = cms.untracked.int32(9),
    dataset = cms.untracked.PSet(
        dataTier = cms.untracked.string('LHE'),
        filterName = cms.untracked.string('')
    ),
    eventAutoFlushCompressedSize = cms.untracked.int32(20971520),
    fileName = cms.untracked.string('file:step1_inRAWSIM.root'),
    outputCommands = process.RAWSIMEventContent.outputCommands,
    splitLevel = cms.untracked.int32(0)
)
```

Dominic-Stafford replied:
OK, thanks for explaining this, I've added another commit to swap the datatier order

srimanob commented on Configuration/PyReleaseValidation/python/relval_extendedgen.py:
```diff
@@ -41,6 +41,8 @@
 workflows[519]=['WTolNuJets_LO_Mad_13TeV_py8_taurhonu',['WTolNu01234Jets_5f_LO_MLM_Madgraph_LHE_13TeV','Hadronizer_TuneCUETP8M1_13TeV_MLM_5f_max4j_LHE_pythia8_taurhonu','HARVESTGEN2']]
 workflows[520]=['VHToHtt_NLO_Pow_13TeV_py8_taurhonu',['VHToH_Pow_LHE_13TeV','Hadronizer_TuneCUETP8M1_13TeV_Httrhonu_powhegEmissionVeto_pythia8','HARVESTGEN2']]
 
+workflows[535] = ['', ['TTbar_13TeV_Pow_herwig7','HARVESTGEN']]
```
This workflow is not in the master. What is the plan for master?

chayanit replied:
Hi @srimanob 535 is there in master https://github.com/cms-sw/cmssw/blob/master/Configuration/PyReleaseValidation/python/relval_extendedgen.py#L26

srimanob replied:
Thanks. Can it move to the place where it is in master? The code is very confusing to compare now. Order is confusion, and it is annoying when I try to compare with master.

Dominic-Stafford replied:
I put this at the end of the "LHE Generation + Hadronization" block since this is what this workflow is. I'm not sure why in master it's in a "Matrix Element Generations" block, but when I make a PR to master to add more workflows, I could move it to where this one is now?

srimanob replied:
Thanks, please agree with @cms-sw/generators-l2 and adapt it to the proper place.

agrohsje replied:
I agree with LHE + Hadronization. We should fix in master.

srimanob commented:
test parameters:

- workflow = 535
- what = extendedgen
- pull_request = https://github.com/cms-sw/cmsdist/pull/6829

srimanob commented:
test parameters:

- workflow = 535
- relvals_options = --what extendedgen
- pull_request = https://github.com/cms-sw/cmsdist/pull/6829

srimanob commented:
please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-947a95/14694/summary.html
**COMMIT**: 998b96f1771cc3041ed14075cf71fff335e064e0
**CMSSW**: CMSSW_9_4_X_2021-04-25-0000/slc6_amd64_gcc630
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33492/14694/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-947a95/535.0_TTbar_13TeV_Pow_herwig7+TTbar_13TeV_Pow_herwig7+HARVESTGEN

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files

Dominic-Stafford committed:
Change order of datatier in relval

cmsbuild commented:
Pull request #33492 was updated. @SiewYan, @mkirsano, @jordan-martins, @chayanit, @wajidalikhan, @kpedro88, @cmsbuild, @GurpreetSinghChahal, @srimanob, @agrohsje, @alberto-sanchez can you please check and sign again.

srimanob commented:
Please test

cmsbuild commented:
+1

**Summary**: https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-947a95/14701/summary.html
**COMMIT**: 223f443dff760b26d5877e0fc8974ff87ba83442
**CMSSW**: CMSSW_9_4_X_2021-04-25-0000/slc6_amd64_gcc630
**User test area**: For local testing, you can use `/cvmfs/cms-ci.cern.ch/week0/cms-sw/cmssw/33492/14701/install.sh` to create a dev area with all the needed externals and cmssw changes.


## Comparison Summary



@slava77 comparisons for the following workflows were not done due to missing matrix map:
   * /data/cmsbld/jenkins/workspace/compare-root-files-short-matrix/data/PR-947a95/535.0_TTbar_13TeV_Pow_herwig7+TTbar_13TeV_Pow_herwig7+HARVESTGEN

**Summary**:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 1
 - DQMHistoTests: Total histograms compared: 0
 - DQMHistoTests: Total failures: 0
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 0
 - DQMHistoTests: Total skipped: 0
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0 KiB( 0 files compared)
 - Checked 0 log files, 0 edm output root files, 1 DQM output files

srimanob commented:
+Upgrade

This PR is to upgrade the Herwig generator. A new workflow in GEN validation is added. PR test passes.
@cms-sw/generators-l2 @cms-sw/pdmv-l2 Could you please consider to re-sign it again? The PR was updated to correct the eventcontent - datatier.

silviodonato commented:
merge
